### PR TITLE
feat(core): Add `AbortSignal` support.

### LINKED ISSUE #8853: Support for `AbortSignal` in all method calls
### Which package is the feature request for?

discord.js

### Feature

`@discordjs/rest` supports a `signal` option which lets the user abort requests before or while they happen, however, `discord.js` does not expose such option at the moment.

### Ideal solution or implementation

To start with, this would require a library-wide refactor to add a new option, however, it cannot be done seamlessly and in a consistent manner for v14 because some options have `reason` within the body ([`BanOptions.reason`](https://discord.js.org/#/docs/discord.js/main/typedef/BanOptions)), and some outside as a single parameter ([`GuildMember#kick`](https://discord.js.org/#/docs/discord.js/main/class/GuildMember?scrollTo=kick)).

While the former is straightforward as the change would only require adding one extra parameter, the latter would require combined type checks with likely deprecation warnings for the current version. So this issue will most likely be handled for v15.

Extending the scope a little more, we will very likely switch to `@discordjs/core` (see #8736) for v15, which adds the requirement for `/core` to support this option so the main library can access to this feature. Keep in mind that this won't be done in the initial PR, but in a later PR.

### Alternative solutions or implementations

Supporting `signal` in v14 despite the feature possibly arising inconsistencies in the current library API.

### Other context

PR that added support for `signal` in REST: #8672

~~*Note for near future me: please add the `packages:core` label once it's created*~~

### PR DESCRIPTION
**Pull request by suneettipirneni**
**Please describe the changes this PR makes and why it should be merged:**

Adds an options object to all functions that allows an `AbortSignal` to be passed in. 

In terms of refactoring, all JSON body payloads are now called `body`. Any query parameters object are now called `query`. `reason` parameters are now put in the same option object as `signal`. As a result, this is breaking.

Closes #8853

**Status and versioning classification:**

- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)
- This PR includes breaking changes (methods removed or renamed, parameters moved or removed)

### CODE REVIEW

File: packages/core/src/api/applicationCommands.ts
Comment (kyranet) [inline packages/core/src/api/applicationCommands.ts:251]: This needs to default, otherwise `bulkOverwriteGuildCommands(APP_ID, GUILD_ID, body)` will throw an error as it cannot deconstruct `signal` from `undefined`.

Also, without the `= {}`, this argument is required.

File: packages/core/src/api/interactions.ts
Comment (kyranet) [inline packages/core/src/api/interactions.ts:140]: You shouldn't destructure signal and create a new object, just pass it through

File: packages/core/src/api/invite.ts
Comment (kyranet) [inline packages/core/src/api/invite.ts:34]: Double `undefined` in reason.

**Review by kyranet — COMMENTED** [review]
Partial review with wide scope — also, can't we centralize those types? We can also use `Pick<RequestData, 'signal'>`.

File: packages/core/src/api/roleConnections.ts
Comment (kyranet) [inline packages/core/src/api/roleConnections.ts:36]: This (`options`) parameter is not documented.

File: packages/core/src/api/thread.ts
Comment (kyranet) [inline packages/core/src/api/thread.ts:46]: For a follow-up PR, this should be separated in two — `messageId: Snowflake` and `body: Omit<StartThreadOptions, 'message_id'>`.

File: packages/core/src/api/webhook.ts
Comment (kyranet) [inline packages/core/src/api/webhook.ts:68]: We can probably use `Pick<> & { token?: string | undefined }` here, likewise everywhere else in this file.

Otherwise, `token?: string` → `token?: string | undefined`.

File: packages/core/src/api/interactions.ts
Comment (kyranet) [inline packages/core/src/api/interactions.ts:92]: `options:` fields don't need defaulting to an empty object if they're not accessed.

File: packages/core/src/api/webhook.ts
Comment (kyranet) [inline packages/core/src/api/webhook.ts:105]:
```suggestion
		options?: Pick<RequestData, 'signal'>,
```

File: packages/core/src/api/webhook.ts
Comment (kyranet) [inline packages/core/src/api/webhook.ts:121]:
```suggestion
		options?: Pick<RequestData, 'signal'>,
```

**Review by kyranet — CHANGES_REQUESTED** [review]
I think we should decide which pattern to use for the options:

1. Current (except one occurrence, I don't find it anymore):
	```typescript
	async foo({ signal }: Pick<RequestData, 'signal'> = {}) {
		await this.rest.bar(Routes.baz(), { signal });
	}

	async foo(body: HelloWorld, { signal }: Pick<RequestData, 'signal'> = {}) {
		await this.rest.bar(Routes.baz(), { body, signal });
	}
	```
1. Spread
	```typescript
	async foo(options?: Pick<RequestData, 'signal'>) {
		await this.rest.bar(Routes.baz(), options);
	}

	async foo(body: HelloWorld, options?: Pick<RequestData, 'signal'>) {
		await this.rest.bar(Routes.baz(), { ...options, body });
	}
	```

The upside of the "Spread" approach is that it doesn't allocate extra objects if none is passed, and also removes a lot of visual clutter while keeping the code shorter.

**Review by kyranet — APPROVED** [review]

File: packages/core/src/api/applicationCommands.ts
Comment (vladfrangu) [inline packages/core/src/api/applicationCommands.ts:73]: You made it default to an empty object twice and then never again 👁️

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: packages/core/src/api/webhook.ts
Comment (Jiralite) [inline packages/core/src/api/webhook.ts:241]:
```suggestion
		{ signal }: { signal?: AbortSignal | undefined } = {},
```

File: packages/core/src/api/webhook.ts
Comment (vladfrangu) [inline packages/core/src/api/webhook.ts:67]: You renamed all to body

File: packages/core/src/api/webhook.ts
Comment (vladfrangu) [inline packages/core/src/api/webhook.ts:216]: Isn't there a -types interface for this?

File: packages/core/src/api/webhook.ts
Comment (vladfrangu) [inline packages/core/src/api/webhook.ts:241]: Why isn't this consistent with the rest (pick<>)

File: packages/core/src/api/guild.ts
Comment (vladfrangu) [inline packages/core/src/api/guild.ts:123]: Didn't you rename all bodies to body

File: packages/core/src/api/guild.ts
Comment (vladfrangu) [inline packages/core/src/api/guild.ts:152]: Doesn't match signature anymore

File: packages/core/src/api/guild.ts
Comment (vladfrangu) [inline packages/core/src/api/guild.ts:157]: Consistency (Pick<>)

File: packages/core/src/api/guild.ts
Comment (vladfrangu) [inline packages/core/src/api/guild.ts:622]: Pick<>

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: packages/core/src/api/webhook.ts
Comment (suneettipirneni) [reply packages/core/src/api/webhook.ts:216]: Doesn't seem to be one.

File: packages/core/src/api/guild.ts
Comment (vladfrangu) [inline packages/core/src/api/guild.ts:141]: isn't it supposed to be body?

File: packages/core/src/api/interactions.ts
Comment (vladfrangu) [inline packages/core/src/api/interactions.ts:159]: Consistency with destructuring please

File: packages/core/src/api/oauth2.ts
Comment (vladfrangu) [inline packages/core/src/api/oauth2.ts:61]: Consistency with naming

File: packages/core/src/api/oauth2.ts
Comment (vladfrangu) [inline packages/core/src/api/oauth2.ts:84]: Ditto

File: packages/core/src/api/user.ts
Comment (vladfrangu) [inline packages/core/src/api/user.ts:100]: This fits better under guild as `editCurrentGuildMember`, the same reason for the voice state one 👀

File: packages/core/src/api/webhook.ts
Comment (vladfrangu) [inline packages/core/src/api/webhook.ts:105]: Destructuring

File: packages/core/src/api/webhook.ts
Comment (vladfrangu) [inline packages/core/src/api/webhook.ts:121]: Ditto

File: packages/core/src/api/webhook.ts
Comment (vladfrangu) [reply packages/core/src/api/webhook.ts:216]: Lets make one then!

**Review by vladfrangu — CHANGES_REQUESTED** [review]

**Review by vladfrangu — APPROVED** [review]
LGTM after a rebase

**Review by Jiralite — APPROVED** [review]

**Review by iCrawl — APPROVED** [review]

**Pull request merged**
